[RFC] LwM2M: write attribute support

rbtchc opened the pull request:
This is a proposal to support [LwM2M attributes](http://www.openmobilealliance.org/release/LightweightM2M/V1_0_1-20170704-A/OMA-TS-LightweightM2M-V1_0_1-20170704-A.pdf), sec 5.1

Because attributes can be associated with object, object instance, resource, I decided to add a struct lwm2m_attr which is a generic container for any type of attribute and a "sys_slist_t" to the obj/obj_inst/res structure for holding the attributes.

Also, add implementation for
1. lwm2m_write_attr_handler(): check attributes written by server, convert and store into lwm2m_attr in requested obj/obj_inst/res. Update the **observe_node** affected w/ the write_attr op accordingly
1. do_discover_op(): append attributes to the requested link format
1. engine_add_observer(): update to get pmin/pmax from the stored attributes if any

PR #4981, #5099 are included in this PR but shall be reviewed separately.

Fixes #3856

rbtchc committed:
net: lwm2m: add write-attribute WRITE support

Implement write-attribute on obj/obj_inst/res according to LwM2M spec
20170704-A, sec 5.1.2. Support pmin/pmax/st/gt/lt parameters on WRITE
operation.

The basic idea is to add sys_slist_t to obj/obj_inst/res structure.
And attach struct lwm2m_attr to the list when attributes are written
from server side (implement lwm2m_write_attr_handler accordingly)

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: lwm2m: report attributes on discover op

Since we've added storing notification attributes written by server.
We can now append these attributes as part of link-format for discover
op.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: lwm2m: get attributes from obj/obj_inst/res for observe_node

1. Read notification attributes set by server to setup the
   minimum/maximum notification period of a observation request.
2. Reordering to check observe_node duplication first
   (bailout earlier)
3. Simplify remove observe_node condition checking

NOTE: attributes are inheritable, priority: res > obj_inst > obj
Reference: LwM2M spec V1_0_1-20170704-A, section 5.1

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

lpereira commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2261,18 +2571,62 @@ static int do_read_op(struct lwm2m_engine_obj *obj,
+
+static int do_discover_op(struct lwm2m_engine_context *context, bool well_known)
 {
 	static char disc_buf[24];
```
Unrelated to the patch, but is there a way that do_discover_op() can be called from two different threads?

lpereira commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -699,6 +784,71 @@ static u16_t atou16(u8_t *buf, u16_t buflen, u16_t *len)
 	return val;
 }
 
+static int atos32(u8_t *buf, u16_t buflen, s32_t *val)
```
There's a strtol() in the minimal libc. Can that be used instead? (Should save a few hundred bytes.)

rbtchc replied:
Yes, good suggestion. I replace atos32() w/ strtol().

lpereira commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -699,6 +784,71 @@ static u16_t atou16(u8_t *buf, u16_t buflen, u16_t *len)
+	return buflen == pos ? 0 : -EINVAL;
+}
+
+static int atof32(u8_t *buf, u16_t buflen, float32_value_t *val)
```
Is this a simplified floating pointing notation without exponents, NaN, inf, et. al.?

rbtchc replied:
Yes, I believe so.
Detail is described at src/zephyr/include/net/lwm2m.h, line 137~152.

I think it should be sufficient for lwm2m usage.

mike-scott reviewed (commented):
Added review for the 2 open PRs which this includes.  I'll do a full review of this PR once those are merged and this is rebased.

I like the overall approach.

rbtchc commented:
Update to
1. replace atos32() w/ strtol() function
1. limit pmin/pmax to u16_t for negative are non-sense

Diff: https://hastebin.com/efixiyagek.cs

rbtchc commented:
Minor tweaks. diff: https://hastebin.com/sapawutebo.php
1. engine_add_observer(): check observe_node duplication first to bail out earlier
1. engine_remove_observer_by_id(): simplify condition checking
1. lwm2m_write_attr_handler(): hide security object;

lpereira commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -699,6 +797,45 @@ static u16_t atou16(u8_t *buf, u16_t buflen, u16_t *len)
+	}
+
+	val = strtol(buf, &end, 10);
+	if (*end || val > INT_MAX || val < INT_MIN) {
```
LONG_MIN and INT_MIN are usually the same on 32-bit (at least that seems to be the case in the minimal libc); same goes for the MAX values.  So it's better to check for errno here:

```c
errno = 0;
val = strtol(buf, &end, 10);
if (errno || *end) {
    return -EINVAL;
}
```

(Zeroing out `errno` before calling `strtol()` is necessary since it does not reset if no error was detected.)

rbtchc commented:
Updated w/ following changes
1. Address the range issue pointed out by @lpereira
1. Update the existing "observe_node" when attributes are written from the server side

Diff: https://hastebin.com/folasacixa.cs

mike-scott commented:
This will need a rebase once https://github.com/zephyrproject-rtos/zephyr/pull/4981 and https://github.com/zephyrproject-rtos/zephyr/pull/5099 are merged.

rbtchc committed:
net: lwm2m: update observe_node when attributes are written or cleared

As title, we should update the existing observe_node when new attributes
are written from server side. Add the implementation to handle that.

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc commented:
Updated w/ following changes
1. Correct atof32 errors
    - storing fraction w/ sign for comparison
    - checking sign from the input string. you cannot tell sign by checking the decimal part (e.g. -0.1, decimal = 0)
1. Modify to store the attributes in **struct notification_attrs** first and do the validation before we apply the changes
    - check pmin < pmax
    - check lt < gt
    - check lt + 2*st < gt

Diff: https://hastebin.com/axosakulok.scala

rbtchc commented:
No diff. Rebased on top of master branch

rbtchc commented:
Update: check step only when it's set

Diff: https://hastebin.com/ajefagabav.diff

rbtchc commented:
Rebase on top of master branch

mike-scott reviewed (approved):
LGTM.  Tested w/ Wakaama server setting attributes, watching observer timings change, clear attributes while still observing and it reflected the changes accordingly etc.  Checked discovery to make sure attributes were being rendered correctly.

mike-scott commented:
@lpereira did you have any further issues with this PR?  It's probably good to go.
